### DISCUSSION TITLE
Unable to Build and Run Code Server Code in Windows Laptop

### DISCUSSION THREAD

**Discussion by sandeepgulecha** [coder/code-server#6957] [Q&A]
I have downloaded source code of code-server from github. Installed all the required packages in Windows Laptop. When I tried to run command yarn run build command in git bash, out folder is not getting generated. What could be issue? Is there any doc to build and run in local window OS laptop.

**Comment by code-asher** [comment]
At the moment running on Windows is not supported but PRs to make it work are definitely welcome.  https://github.com/coder/code-server/issues/1397

**Reply by sandeepgulecha** [reply]
Is it possible build in Mac Laptop local?

**Reply by code-asher** [reply]
Yup, macOS should work.

**Reply by sandeepgulecha** [reply]
I am able to build and run code server code in macOS local. Under lib/vscode folder, unable to build it and generate out folder. Can you please provide steps to build and generate out folder for vscode folder code.  After change file app.ts in code-server/src/node folder able to build and generate the out folder but while run the application using url 127.0.0.1:8080 get the following error mod.loadCodeWithNls is not a function. how to fix this error.
![image001](https://github.com/user-attachments/assets/dac86c9c-d5ab-45ba-ac6c-2085369bbe0c)
<img width="1228" alt="Screenshot 2024-08-29 at 5 58 03 PM" src="https://github.com/user-attachments/assets/4e639dc8-6b66-4976-a1d0-6adf8f5c7813">

**Reply by code-asher** [reply]
The steps are all in https://github.com/coder/code-server/blob/main/docs/CONTRIBUTING.md, but the gist is that you need to make sure you have all the dependencies installed then run:

```
git clone https://github.com/coder/code-server.git
cd code-server
git submodule update --init
quilt push -a
yarn 
yarn build
VERSION=0.0.0 yarn build:vscode
yarn release
```

From there you can run the release with `node ./release`.

**Reply by sandeepgulecha** [reply]
Thanks a lot. Its works. I am able to build and run in macOS after code modification in code server ts file. I need to run the application in WindowOS which is build in macOS. In WindowsOS Laptop, I have installed code-server from npm. When I copied code-server out folder from mac to windowOS system and getting same error. What is steps to make work in WindowOS which is build in macOS. What is vscode out folder need to copied from macOS to WindowOS?
![image001](https://github.com/user-attach0ments/assets/5fb4b25f-b50a-4e95-a2f3-0fd80c689dd6)<img width="1228" alt="Screenshot 2024-08-29 at 5 58 03 PM" src="https://github.com/user-attachments/assets/2739ad6a-fdde-4efe-9bc2-2881536853e4">

**Reply by code-asher** [reply]
From the release directory, `lib/vscode/out` is the out directory for VS Code, but it looks like you already copied it?  I am not sure what that error means.  What version of Node are you using to run this?  Is it v20?

**Reply by code-asher** [reply]
Actually it looks like the patches were not applied.  With patches applied you would not see `Server bound to`

**Reply by sandeepgulecha** [reply]
What is command/steps to apply patches? Can we schedule meeting for 15-30 mints to fix? How I schedule meeting?

**Reply by sandeepgulecha** [reply]
We want to redirect from Application with URL for ex 'https://localhost:4200/' by passing Authentication token in header
 to Code Server URL http://127.0.0.1:8080/ in new tab browser. We have enabled CORS with specific options in Code Server app.ts file but Code Server application response header does not return with Access-Control-Allow-Origin: * or http://127.0.0.1:8080/. how to set Access-Control-Allow-Origin in response header so that it will application to redirect to code server URL http://127.0.0.1:8080/ in new tab browser  

Added following code in app.ts and also attached file (app.txt) as well
// Enable CORS with specific options
const corsOptions = {
  origin: 'https://localhost:4200/', // Replace with the domain of the application making the request
  methods: ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'],
  allowedHeaders: ['Authorization', 'Content-Type'],
  credentials: true,
  preflightContinue: false,
  optionsSuccessStatus: 204  // Respond with 204 for preflight requests
};
// Apply the CORS middleware
app.use(cors(corsOptions));

// Add middleware to handle CORS preflight requests and avoid redirecting
app.options('*', cors(corsOptions));  // Enable CORS pre-flight for all routes

![Network Error 1](https://github.com/user-attachments/assets/212bbe4d-7a29-4865-9717-fd60e26f19c0)
![Network Error 2](https://github.com/user-attachments/assets/ff096d0a-54ea-4b59-ad82-8a384c52eb6e)
[app.txt](https://github.com/user-attachments/files/16852712/app.txt)

**Reply by code-asher** [reply]
The command to apply patches is in the contributing documentation, but it is `quilt apply -a`.

Are you using the Express CORS middleware?  You might have to check with them, it looks to me like you are using it correctly.  Can you confirm their middleware is being called?  Maybe they have a debug flag?

**Reply by sandeepgulecha** [reply]
We have created one custom extension and installed it in code-server in our local with activate command name “code-server-analytics.helloWorld” 

 

Sample code from custom extension’s package.json

"contributes": {
    "commands": [
      {
        "command": "code-server-analytics.helloWorld",
        "title": "Hello World"
      }
    ]
  },


We need to activate this extension using router get() by passing parameter to extension from app.ts  in code-server.
Could you help us.

 We tried below code in our app.ts file in code-server but, we are getting the error
“Warning: 'command' is not in the list of known options, but still passed to Electron/Chromium when running on mac terminal”

router.get("/trigger-extension-action", (req, res) => {
    // For example, run a shell command or trigger some action
    exec("code --command code-server-analytics.helloWorld parameter1", (err, stdout, stderr) => {
      console.log("stdout: ", stdout)
      console.log("stderr: ", stderr)
      if (err) {
        console.log("err: ", err)
        res.status(500).send(stderr)
        return

      }
      console.log("stdout2: ", stdout)
      res.send(stdout)
    })

  })

**Reply by code-asher** [reply]
Based on the error it looks like VS Code does not have a `--command` flag, so that will not work.

**Reply by code-asher** [reply]
Oh unless `--command` is something you added?  In which case it looks like you are running VS Code directly, not code-server, so maybe you just need to run `code-server --command`.

**Reply by sandeepgulecha** [reply]
Code Server is already running, Created Extension in VSCode that will open the file and show file content in vscode.workspace.openTextDocument(filepath).then((doc)  => {vscode.window.showTextDocument(doc)})  statement. Need to call the extension when required. Please let me know what need to written.

**Comment by sandeepgulecha** [comment]
If we write code-server --command does it create new instance of server & activate command or it will call activate command and it will not create new server


Or below code will work without code-server --command 

In extension.ts

 

import * as vscode from 'vscode';

import * as express from 'express';

 

export function activate(context: vscode.ExtensionContext) {

    // Create a new Express server

    const app = express();

    app.use(express.json());  // To handle JSON request bodies

 

    // Route to handle incoming HTTP requests

    [app.post](http://app.post/)('/open-file', async (req, res) => {

        const token = req.body.token;

        const filePath = req.body.filePath;

 

        // Use the token (for authentication) if needed

        console.log('Received Token:', token);

 

        try {

            // Open the file in the VSCode editor

            const document = await vscode.workspace.openTextDocument(filePath);

            await vscode.window.showTextDocument(document);

 

            res.status(200).send({ message: 'File opened successfully' });

        } catch (error) {

            console.error('Error opening file:', error);

            res.status(500).send({ error: 'Failed to open the file' });

        }

    });

 

    // Start the server

    const port = 3000;  // Define a port for the server

    app.listen(port, () => {

        console.log(`VSCode extension HTTP server running on [http://localhost:${port}`](http://localhost:$%7bport%7d%60/));

    });

 

    // Register a sample command (optional)

    let disposable = vscode.commands.registerCommand('extension.helloWorld', () => {

        vscode.window.showInformationMessage('Hello from VSCode extension HTTP Server!');

    });

 

    context.subscriptions.push(disposable);

}

 

export function deactivate() {}

 

 

 

In app.ts file

 

import axios from 'axios';

import express from 'express';

 

const app = express();

 

app.use(express.json());  // To parse JSON request bodies

 

// Sample route in app.ts to trigger the request to the extension's HTTP server

[app.post](http://app.post/)('/trigger-extension', async (req, res) => {

    const token = req.body.token;

    const filePath = req.body.filePath;

 

    try {

        // Send a request to the extension's HTTP server running on port 3000

        const response = await [axios.post](http://axios.post/)('http://localhost:3000/open-file', {

            token: token,  // Pass the token

            filePath: filePath  // Pass the file path

        });

 

        res.status(200).send({ message: 'File opened successfully in VSCode via the extension' });

    } catch (error) {

        console.error('Error calling extension HTTP server:', error);

        res.status(500).send({ error: 'Failed to trigger the extension' });

    }

});

 

// Start the app.ts server on port 8080

app.listen(8080, () => {

    console.log('App server running on http://localhost:8080/');

});

**Reply by code-asher** [reply]
`--command` does not exist, unless you added it, so I am not sure what it is supposed to do.

```
$ code-server --command
[2024-09-11T21:26:22.888Z] error Unknown option --command
```

**Reply by sandeepgulecha** [reply]
How to activate custom created  extension code-server-analytics.helloWorld & installed in app.ts file from code server?

Is it possible to activate extension in app.ts file? If yes, how?

Also I have created new extension in WindowsOS and installed extension.vsix but I am not able to activate it using command palette. What is issue?
Created new extension in macOS and installed extension.vsix and I am able to activate it using command palette

**Reply by code-asher** [reply]
I have no idea why it would not work in Windows, you may need to attach a debugger or add some logs to see if your extension is being called.  If it is not being called, check the extension host logs to see whether the extension was loaded and activated and if there are any errors.

As for launching an extension externally, the only way I can think of is to use a URI handler.  See `registerUriHandler` in https://code.visualstudio.com/api/references/vscode-api.  You need the `onUri` activation event in your `package.json` to use it.

But, unfortunately I do not think code-server actually supports this at the moment, so there might be no way to do this.
